core/storage/couchStorageHelpers.go

```diff
@@ -0,0 +1,582 @@
+package storage
+
+import (
+	"context"
+	"fmt"
+	_ "github.com/go-kivik/couchdb/v3" // The CouchDB Driver
+	kivik "github.com/go-kivik/kivik/v3"
+	"github.com/open-horizon/edge-sync-service/common"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/open-horizon/edge-utilities/logger"
+	"github.com/open-horizon/edge-utilities/logger/log"
+	"github.com/open-horizon/edge-utilities/logger/trace"
+)
+
+func (store *CouchStorage) checkObjects() {
+	if !store.connected {
+		return
+	}
+
+	var err error
+	currentTime := time.Now().UTC().Format(time.RFC3339)
+	query := map[string]interface{}{
+		"selector": map[string]interface{}{
+			"$and": []map[string]interface{}{
+				map[string]interface{}{"metadata.expiration": map[string]interface{}{"$ne": ""}},
+				map[string]interface{}{"metadata.expiration": map[string]interface{}{"$lte": currentTime}},
+				map[string]interface{}{"$or": []map[string]interface{}{
+					map[string]interface{}{"status": common.NotReadyToSend},
+					map[string]interface{}{"status": common.ReadyToSend}}}},
+		},
+		"fields": []string{"metadata", "last-update"},
+	}
+
+	result := []couchObject{}
+	if err := store.findAll(query, &result); err != nil {
+		if err != notFound && log.IsLogging(logger.ERROR) {
+			log.Error("Error in CouchStorage.checkObjects: failed to remove expired objects. Error: %s\n", err)
+		}
+		return
+	}
+
+	if trace.IsLogging(logger.TRACE) {
+		trace.Trace("Removing expired objects")
+	}
+
+	for _, object := range result {
+
+		id := createObjectCollectionID(object.MetaData.DestOrgID, object.MetaData.ObjectType, object.MetaData.ObjectID)
+		if trace.IsLogging(logger.TRACE) {
+			trace.Trace("Deleting object %s\n", id)
+		}
+
+		var query interface{}
+		query = map[string]interface{}{"selector": map[string]interface{}{"_id": id}}
+		if !object.LastUpdate.IsZero() {
```

Comment:
I saw these checks being done in the Mongo implementation, as per my understanding, this is to check if the 'LastUpdate' value is not set, in which case it would be the default value. This is checked using 'LastUpdate.IsZero()'. It seemed like it is checked to ensure that the object has not been updated since it was last queried, because 'RemoveAll()' wasn't directly called. But this is a good point, upon reconsideration, I am not sure if these are needed because the object is removed regardless..could you confirm?

Reply:
ah, I think I figured it out. Since we are deleting expired objects, an update to the object between the query and actual deletion COULD have been an update to the expiration time, which means the deletion should not occur right now. I presume it will occur later. I think the code is correct as you have it.